### PR TITLE
added DirConstantFixer: transforms dirname(__FILE__) into __DIR__

### PR DESCRIPTION
**Pull request by kalessil**

### CODE REVIEW

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                }
+            }
+
+            // get rid of root namespace when it used
+            if ($prevToken->isGivenKind(T_NS_SEPARATOR)) {
+                $matches[0] = $prevTokenIndex;
+            }
+
+            // transform construction
+            $tokens->overrideRange($matches[0], $matches[3], array(new Token(array(T_DIR, '__DIR__'))));
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: this will reduce this amount of tokens (if I read this correct), should/could the `$end` be updated here as well (ie. decrease it with two?)

:+1:  for writing this fixer, great one! :)

(maybe I should work on this after all https://github.com/composer/composer/pull/4220 ;) )

Comment (kalessil) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: Right, I missed that tiny detail...
Thank you, hope that PR will be accepted too )

File: Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            array('<?php "SELECT ... dirname(__FILE__) ...";'),
+            array('<?php "SELECT ... DIRNAME(__FILE__) ...";'),
+            array('<?php "test" . "dirname" . "in concatenation";'),
+
+            array('<?php $x = __DIR__;', '<?php $x = dirname(__FILE__);'),
+            array('<?php $x = __DIR__;', '<?php $x = \\dirname(__FILE__);'),
+            array('<?php $x = __DIR__.".dist";', '<?php $x = dirname(__FILE__).".dist";'),
+            array('<?php $x = __DIR__.".dist";', '<?php $x = \\dirname(__FILE__).".dist";'),
+
+            array($multiLinePatternFixed, $multiLinePatternToFix),
```
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php]: could you add `$a=/**/dirname/**/  /** a*/(/**//** */__FILE__/***/)/*aa*/;` ? I know, insane PHP but valid code...

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            $prevToken = $tokens[$prevTokenIndex];
+            if ($prevToken->isGivenKind(array(T_DOUBLE_COLON, T_NEW, T_OBJECT_OPERATOR, T_FUNCTION))) {
+                continue;
+            }
+
+            // handle function reference with namespaces
+            if ($prevToken->isGivenKind(T_NS_SEPARATOR)) {
+                $twicePrevTokenIndex = $tokens->getPrevMeaningfulToken($prevTokenIndex);
+                $twicePrevToken = $tokens[$twicePrevTokenIndex];
+                if ($twicePrevToken->isGivenKind(array(T_DOUBLE_COLON, T_NEW, T_OBJECT_OPERATOR, T_FUNCTION, T_STRING, CT_NAMESPACE_OPERATOR))) {
```
Comment (gharlan) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: `T_DOUBLE_COLON`, `T_OBJECT_OPERATOR` and `T_FUNCTION` should not be possible here..

```
::\dirname
->\dirname
function \dirname
```

File: Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php (hunk 2 above)
Comment (kalessil) [inline Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php]: Done.

I also have personal collection of horrible code: https://bitbucket.org/kalessil/phpinspectionsea/src/2609331943ef421951a74be3d60bc0fada087a47/php/?at=master

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php (hunk 3 above)
Comment (kalessil) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: Huge thanks, fixed this.

File: Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php [hunk 4]
```diff
@@ -73,7 +73,7 @@ public function fix(\SplFileInfo $file, Tokens $tokens)
             if ($prevToken->isGivenKind(array(T_NS_SEPARATOR))) {
                 $twicePrevTokenIndex = $tokens->getPrevMeaningfulToken($prevTokenIndex);
                 $twicePrevToken = $tokens[$twicePrevTokenIndex];
-                if ($twicePrevToken->isGivenKind(array(T_DOUBLE_COLON, T_NEW, T_OBJECT_OPERATOR, T_FUNCTION, T_STRING, CT_NAMESPACE_OPERATOR))) {
+                if ($twicePrevToken->isGivenKind(array(T_NEW, T_STRING, CT_NAMESPACE_OPERATOR))) {
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php]: ?

File: Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php (hunk 2 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php]: :+1:   thanks!
nice collection :D

File: Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php (hunk 4 above)
Comment (kalessil) [inline Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php]: @gharlan noticed some senseless checks, so I cleaned them up in my another fixer as well.

Comment (SpacePossum) [inline Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php]: Makes the PR less atomic, but seems legit to me. Tests are green as well so :+1:  :)

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php]: This needs to be a separate PR IMO. Ping @keradus.

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: `overrideRange` will **NOT** reduce collection size, never. It may just enlarge it (but it is not a case here)

File: Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php (hunk 4 above)
Comment (keradus) [inline Symfony/CS/Fixer/Symfony/AliasFunctionsFixer.php]: It's always great to have atomic PRs, but here - it's just one line...

File: Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            array('<?php $x = /**/__DIR__/*xx*/;', '<?php $x = /**/dirname/**/ /** x*/(/**//** */__FILE__/***/)/*xx*/;'),
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php]: sth goes wrong here. let have one simplier example here:
```php
array(
    '<?php __DIR__ //we really need it
    ;',
    '<?php dirname(
        __FILE__ // we really need it
    );', 
),
`

we must not lose that comment

Comment (kalessil) [inline Symfony/CS/Tests/Fixer/Contrib/DirConstantFixerTest.php]: Okay, I'll need to change logic then.

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: I see, but are you sure? This is then confusing to me:
```php
        // If we want to add less tokens than passed range contains then clear
        // not needed tokens.
```

Comment (keradus) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: Yes, I'm sure. Clearing the token made it empty, not remove it from collection.

Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: ah ok, I misread, :+1:

File: Symfony/CS/Fixer/Contrib/DirConstantFixer.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            // stop looping if didn't find any new matches
+            if (null === $matches) {
+                break;
+            }
+
+            // 0 to 3 accordingly are "dirname", "(", "__FILE__", ")"
+            $matches = array_keys($matches);
+
+            // move cursor just after sequence
+            $currIndex = $matches[3];
```
Comment (stof) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: This is still wrong, as you can have more than 4 tokens in the sequence (because if spaces and comments)

Comment (kalessil) [inline Symfony/CS/Fixer/Contrib/DirConstantFixer.php]: Yep, was noticed upper - I need to adjust logic to preserve tokens inside parameters list (originally comments, but space as well).

**Comment by gharlan** [comment]
IMO this is a candidate for symfony level.

**Comment by keradus** [comment]
yes it is

**Comment by GrahamCampbell** [comment]
> yes it is

It can't be because of the fact it doesn't check if the function is defined in a namespace.

**Comment by keradus** [comment]
We already check if it call's with namespace (ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1302/files#diff-492ffd9b1d024084a7238cb0e9678943R55 )

Maybe we should just add a check if function is defined in file and not imported (that just in php7)?
If we don't want to do this then the fixer should be marked as risky.

**Comment by gharlan** [comment]
> Maybe we should just add a check if function is defined in file and not imported (that just in php7)?

The function could be declared in another file but in same namespace. https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1088#discussion_r27161938

**Comment by gharlan** [comment]
> It can't be because of the fact it doesn't check if the function is defined in a namespace.

Then we have to move the alias_function fixer to contrib level, too.

**Comment by GrahamCampbell** [comment]
I thought the alias_function one was contrib, or did it end up getting moved?

**Comment by GrahamCampbell** [comment]
I thought the rule was no risky fixers allowed in symfony?

**Comment by GrahamCampbell** [comment]
This fixer and the alias one are risky.

**Comment by keradus** [comment]
Then both fixers should be marked as risky.
For now Symfony already uses risky fixers under fabbot.io serviece (fixers that are not yet merged about PHPUnit method calls).

**Comment by gharlan** [comment]
> I thought the alias_function one was contrib, or did it end up getting moved?

It is 'symfony'.

> I thought the rule was no risky fixers allowed in symfony?

I don't know. But I vote :+1: for this rule.

So I change my mind and vote :-1: for moving this fixer to symfony level, and :+1: for moving `alias_function` to contrib level.

**Comment by keradus** [comment]
:-1: for moving `alias_function` to contrib level (it's already running on Symfony project), then
:+1: for marking `alias_function` as risky

**Comment by gharlan** [comment]
> it's already running on Symfony project

no, it is added to master branch.

**Comment by keradus** [comment]
on 1.x line the fixer changed just one alias and was named differently:
> join_function [symfony]
> Implode function should be used instead of join function.

**Comment by gharlan** [comment]
After more thought, I would say: If a risky fixer is accepted by symfony deciders it is ok, and the fixer can live in symfony level.

**Comment by keradus** [comment]
Some risky fixers were already requested by them.
That is why we want to add extra check on 2.x line to not run any risky fixers without explicity allowing that in configuration.

**Comment by kalessil** [comment]
@keradus : would please summarize what's to target:

- is_null fixer (risky, Symfony level?) - adjusted
- __DIR__ fixer (risky, Symfony level?) - still need logic changes

**Comment by kalessil** [comment]
Closing, as we've decided to merge this into #1307
